### PR TITLE
Add `--os=all` and `--arch=all` options for `brew audit`.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

~~Needs to be rebased onto https://github.com/Homebrew/brew/pull/15201.~~

### CODE REVIEW

File: Library/Homebrew/dev-cmd/audit.rb [hunk 1]
```diff
@@ -216,59 +239,81 @@ def self.audit
         style_offenses:      style_offenses&.for_path(f.path),
       }.compact
 
-      audit_proc = proc { FormulaAuditor.new(f, **options).tap(&:audit) }
+      oses.product(arches).each do |os, arch|
+        SimulateSystem.with os: os, arch: arch do
+          odebug "Auditing Formula #{f} on os #{os} and arch #{arch}"
+
+          Formulary.clear_cache # TODO: Should cache key include os and arch?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb:246]: Yes, it probably should!

Comment (reitermarkus) [reply Library/Homebrew/dev-cmd/audit.rb]: I don't think it's possible to skip this since the formula class name will stay the same.

File: Library/Homebrew/formulary.rb [hunk 2]
```diff
@@ -661,7 +661,8 @@ def self.factory(
   )
     raise ArgumentError, "Formulae must have a ref!" unless ref
 
-    cache_key = "#{ref}-#{spec}-#{alias_path}-#{from}"
+    os_arch = "#{Homebrew::SimulateSystem.current_os}-#{Homebrew::SimulateSystem.current_arch}"
```
Comment (reitermarkus) [inline Library/Homebrew/formulary.rb:664]: This doesn't seem to change anything. `Formulary.clear_cache` is still needed.

File: Library/Homebrew/cli/args.rb [hunk 3]
```diff
@@ -35,11 +35,13 @@ def freeze_remaining_args!(remaining_args)
...
         self[:named] = NamedArgs.new(
           *named_args.freeze,
-          override_spec: spec(nil),
-          force_bottle:  self[:force_bottle?],
-          flags:         flags_only,
-          cask_options:  cask_options,
-          parent:        self,
+          parent:       self,
+          cask_options: cask_options,
+          **{
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb:40]: Could this be pulled out into variable (and all the parenthesised bits below)? Will make it a bit easier to follow than embedding into params.

File: Library/Homebrew/formulary.rb [hunk 4]
```diff
@@ -311,16 +323,28 @@ def ruby_source_checksum
...
+    klass = T.cast(klass, T.class_of(Formula))
+
+    mod.const_set(class_name, klass)
 
     cache[:api] ||= {}
     cache[:api][name] = klass
   end
 
-  def self.resolve(name, spec: nil, force_bottle: false, flags: [])
+  sig { params(name: String, spec: Symbol, force_bottle: T::Boolean, flags: T::Array[String]).returns(Formula) }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb:334]:
```suggestion
  sig { params(name: String, spec: T.nilable(Symbol), force_bottle: T::Boolean, flags: T::Array[String]).returns(Formula) }
```

no, given these are set to `nil` below?

**Review by MikeMcQuaid — COMMENTED** [review]
This is looking good so far! I have to be honest and say that I haven't given every line detailed review. Is there any way to split this into multiple, smaller PRs for review before merge?

Similarly, I have mild concerns with the amount of code that's not covered by any tests but want to avoid adding new integration tests for all these commands. Suggestions welcome, may not be blocking.

Thanks again @reitermarkus, great work here.

File: Library/Homebrew/formulary.rb (hunk 4 above)
Comment (reitermarkus) [reply Library/Homebrew/formulary.rb:334]: No, the type should be `Symbol`. You should only ever call it with e.g. `spec: :stable`, but never explicitly with `spec: nil`. The `T.unsafe(nil)` is only there to distinguish if the parameter is absent or not.

File: Library/Homebrew/cli/args.rb [hunk 5]
```diff
@@ -33,13 +33,16 @@ def freeze_remaining_args!(remaining_args)
       end
 
       def freeze_named_args!(named_args, cask_options:)
+        options = {
+          override_spec: (:head if self[:HEAD?]),
+          force_bottle:  (true if self[:force_bottle?]),
+          flags:         (flags_only unless flags_only.empty?),
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb:39]: Could these be pulled into variables or `options[:override_spec]` calls outside? The inline `()` usage is a bit hard to follow and not what we typically do.

File: Library/Homebrew/cli/named_args.rb [hunk 6]
```diff
@@ -11,7 +11,24 @@ module CLI
...
+          override_spec: Symbol,
+          force_bottle:  T::Boolean,
+          flags:         T::Array[String],
+          cask_options:  T::Boolean,
+        ).void
+      }
+      def initialize(
+        *args,
+        parent: Args.new,
+        override_spec: T.unsafe(nil),
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/named_args.rb:27]: Should this not be `T.nilable(Symbol)` instead?

File: Library/Homebrew/cmd/fetch.rb [hunk 7]
```diff
@@ -87,52 +88,92 @@ def self.fetch
...
-
-            fetched_bottle = false
-            onoe e.message
-            opoo "Bottle fetch failed, fetching the source instead."
-          else
-            fetched_bottle = true
-          end
-        end
+        formula = T.cast(formula_or_cask, Formula)
+        ref = formula.loaded_from_api? ? formula.full_name : formula.path
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/fetch.rb:98]: This is used in multiple places; maybe worth a dedicated method on `Formula`?

File: Library/Homebrew/extend/os/mac/diagnostic.rb [hunk 8]
```diff
@@ -343,7 +343,7 @@ def check_for_iconv
           nil
         end
         if libiconv&.linked_keg&.directory?
-          unless libiconv.keg_only?
+          unless libiconv&.keg_only?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/mac/diagnostic.rb:346]:
```suggestion
          if !...
```

would be a bit easier to follow

File: Library/Homebrew/formula_versions.rb [hunk 9]
```diff
@@ -38,11 +38,17 @@ def rev_list(branch)
...
   end
 
+  sig { params(rev: String).returns(String) }
   def file_contents_at_revision(rev)
     repository.cd { Utils.popen_read("git", "cat-file", "blob", "#{rev}:#{entry_name}") }
   end
 
-  def formula_at_revision(rev)
+  sig {
+    type_parameters(:U)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_versions.rb:47]: What does this do? Never seen it before. Not blocking!

**Review by MikeMcQuaid — COMMENTED** [review]
Looking good! A few questions and thoughts here but nothing major. Nice work @reitermarkus!

File: Library/Homebrew/cmd/fetch.rb (hunk 7 above)
Comment (reitermarkus) [reply Library/Homebrew/cmd/fetch.rb:98]: I guess we should have something like `Formula#refresh`. I think for casks we actually already have `Cask#refresh`.

File: Library/Homebrew/extend/os/mac/diagnostic.rb (hunk 8 above)
Comment (reitermarkus) [reply Library/Homebrew/extend/os/mac/diagnostic.rb:346]: Pretty sure RuboCop prefers `unless` here.

File: Library/Homebrew/cmd/fetch.rb (hunk 7 above)
Comment (reitermarkus) [reply Library/Homebrew/cmd/fetch.rb:98]: That's significantly harder for formulae though, given the class is already evaluated for a specific system if the formula instance exists.

File: Library/Homebrew/cli/named_args.rb (hunk 6 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cli/named_args.rb:27]: I see. Fine with me!

File: Library/Homebrew/cmd/fetch.rb (hunk 7 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/fetch.rb:98]: Fine with me to leave as-is.

File: Library/Homebrew/extend/os/mac/diagnostic.rb (hunk 8 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/extend/os/mac/diagnostic.rb:346]: Give it a go but ignore me if Rubocop disagrees.

File: Library/Homebrew/formula_versions.rb (hunk 9 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/formula_versions.rb:47]: Cool!

**Review by MikeMcQuaid — APPROVED** [review]
One style nit if you can be bothered but fine if this ships as-is. Thanks again @reitermarkus!

File: Library/Homebrew/extend/os/mac/diagnostic.rb (hunk 8 above)
Comment (reitermarkus) [reply Library/Homebrew/extend/os/mac/diagnostic.rb:346]: Yeah, RuboCop disagrees.

Comment (MikeMcQuaid) [reply Library/Homebrew/extend/os/mac/diagnostic.rb:346]: @reitermarkus Cool, thanks for checking, ship as-is!
